### PR TITLE
[Many] Bump AGP to 8.5.1 and gradle to 8.7

### PR DESCRIPTION
**Pull request by gmackall**
Upgrades all AGP versions less than 8 to latest. The primary motivation is to prepare for https://github.com/flutter/flutter/issues/136879#issuecomment-2292073921, which requires that plugins use a non-7.3 version. We believe it only affects version 7.3 - I'm going to test this, but am not sure the exact range yet and updating is independently valuable, so I just updated all pre 8 versions (it doesn't affect 8+).

Also:
- Upgrades Gradle versions as required
- Marks the use of `Camera2Interop` with `@OptIn`, as it is required by the newer AGP version.
- Changes the Camerax example java directory path to align with the package xml attribute/AGP namespace attribute.
- Sets `android.buildFeatures.buildConfig` to true for compatibility with AGP 8.0+ in some places.
- Sets `exported=true` in a manifest, requires for AGP bump.
- Updates espresso in some examples, required by the AGP bump (0.2.0 isn't compatible with AGP 8.0+).

Also related to https://github.com/flutter/flutter/issues/146660

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

### CODE REVIEW

**Review by tarrinneal — APPROVED** [review]
pigeon, image_picker, shared_preferences lgtm

**Review by stuartmorgan — APPROVED** [review]
LGTM, but please make the PR description and title more descriptive and accurate. Almost none of the packages being touched here appear to be using AGP 7.3, so that's confusing, and then it's also updating Gradle and some miscellaneous other things like the annotations (which I assume are required for the update, so that's fine, but calling them out in the description would still be helpful).

File: packages/camera/camera/example/android/build.gradle [hunk 1]
```diff
@@ -5,7 +5,7 @@ buildscript {
     }
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:7.4.2'
```
Comment (reidbaker) [inline packages/camera/camera/example/android/build.gradle:8]: I understand if you want this in another pr but we also need to migrate to the plugin version of this import.

**Comment by reidbaker** [comment]
> LGTM, but please make the PR description and title more descriptive and accurate. Almost none of the packages being touched here appear to be using AGP 7.3, so that's confusing, and then it's also updating Gradle and some miscellaneous other things like the annotations (which I assume are required for the update, so that's fine, but calling them out in the description would still be helpful).

I updated the pr title but please still add a reason for the @optIn annotation in the pr description.

**Comment by gmackall** [comment]
@stuartmorgan and @reidbaker - updated.

Also, the repo checks are failing for interactive media ads - I'm assuming I should override? I only made changes to the pubspec/gradle files of the example app.

File: packages/camera/camera/example/android/build.gradle (hunk 1 above)
Comment (gmackall) [reply packages/camera/camera/example/android/build.gradle:8]: Yes we should do that at some point, but would prefer for it not to be this pr

**Comment by reidbaker** [comment]
> @stuartmorgan and @reidbaker - updated.
> 
> Also, the repo checks are failing for interactive media ads - I'm assuming I should override? I only made changes to the pubspec/gradle files of the example app.

That is probably a question for @bparrishMines

**Comment by bparrishMines** [comment]
Overriding for `interactive_media_ads` is fine with me since you are only changing the example. We'll need to investigate in the tool why it is failing though.

**Review by bparrishMines — APPROVED** [review]
LGTM

**Comment by bparrishMines** [comment]
@stuartmorgan It looks like `settings.gradle` is missing from this list: https://github.com/flutter/packages/blob/main/script/tool/lib/src/common/package_state_utils.dart#L220

**Comment by stuartmorgan** [comment]
> @stuartmorgan It looks like `settings.gradle` is missing from this list: https://github.com/flutter/packages/blob/main/script/tool/lib/src/common/package_state_utils.dart#L220

Yep, @gmackall Feel free to add that to this PR.

**Comment by gmackall** [comment]
> > @stuartmorgan It looks like `settings.gradle` is missing from this list: https://github.com/flutter/packages/blob/main/script/tool/lib/src/common/package_state_utils.dart#L220
> 
> Yep, @gmackall Feel free to add that to this PR.

Added

**Pull request merged**